File: src/ds-visuals/tree/code/queue.java
class TreeNode{
	int data;
	TreeNode right;
	TreeNode left;
	TreeNode(int data){
		this.data = data;
		this.right = null;
		this.left = null;
	}
}
class Queue{
	final int MAX = 10;
	TreeNode queue[] = new TreeNode[MAX];
	int front = -1;
	int rear = -1;

	void enqueue(TreeNode element){
		if(this.rear == MAX-1)
			System.out.println("Error- queue full");
		else
			this.queue[++rear] = element;
		if(this.rear == 0)
			this.front = 0;
		return;
	}

	TreeNode dequeue(){
		if(this.rear == -1)
			System.out.println("Error- queue empty");
		else{
			TreeNode out = this.queue[front++];
			if(front > rear){
				front = -1;
				rear = -1;
			}
			return out;
		}
		return null;
	}

	void createQueue(TreeNode ... arg){
		for(int i = 0; i < arg.length; i++)
			this.enqueue(arg[i]);
		return;
	}

	void display(){
	    System.out.printf("queue: ");
		for(int i = front; i <= this.rear; i++)
			System.out.printf(this.queue[i]+" ");
		System.out.println();
	}

	boolean isEmpty(){
		if(this.front == -1 || this.front > this.rear)
			return true;
		return false;
	}
}

File: src/ds-visuals/tree/code/queue.c
#include<stdio.h>
#include<stdlib.h>
#include<stdarg.h>
#define MAX 20
//queues using arrays.

typedef struct s{
	int key;
	struct s *right;
	struct s *left;
} TreeNode;

typedef struct q{
	TreeNode* arr[MAX];
	int front;
	int rear;
} Queue;

void enqueue(Queue* queue, TreeNode* element){
	if(queue->rear == MAX-1)
		printf("Error- Queue Full\n");
	else
		queue->arr[++(queue->rear)] = element;
	if(queue->rear == 0)
		queue->front = 0;
	return;	
}

TreeNode* dequeue(Queue*queue){
	if(queue->front == -1 || queue->front > queue -> rear)
		printf("Error- Queue empty\n");
	else{
		TreeNode* out = queue->arr[(queue->front)++];
		if(queue->front > queue->rear){
			queue->front = -1;
			queue->rear = -1;
		}
		return out;	
	}
	return NULL;
}

int isEmpty(Queue queue){
	if(queue.front == -1 || queue.front > queue.rear)
		return 1;
	return 0;
}

void display(Queue *queue){
	int i;
	printf("queue: ");
	for(i = queue->front; i <= queue->rear; i++)
		printf("%d ",queue->arr[i]);
	printf("\n");
	return;
}

Queue createQueue(int num_elements, ...){
	Queue queue;
	int i;
	va_list list;
	queue.front = -1;
	queue.rear = -1;
    va_start(list, num_elements); 
    for (i = 1; i <= num_elements; i++) 
        enqueue(&queue, va_arg(list, TreeNode*)); 
    va_end(list); 
    return queue;  
}